src/jobserver.cc

```diff
@@ -0,0 +1,101 @@
+// Copyright 2024 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "jobserver.h"
+
+#include <cstring>
+#include <cassert>
+
+#include "util.h"
+
+bool Jobserver::Acquire() {
+  // The first token is implicitly handed to the ninja process, so don't
+  // acquire it from the jobserver
+  if (token_count_ == 0 || AcquireToken()) {
+    token_count_++;
+    return true;
+  }
+
+  return false;
+}
+
+void Jobserver::Release() {
+  assert(token_count_ >= 1);
+  token_count_--;
+
+  // Don't return first token to the jobserver, as it is implicitly handed
+  // to the ninja process
+  if (token_count_ > 0) {
+    ReleaseToken();
+  }
+}
+
+bool Jobserver::ParseJobserverAuth(const char* type) {
+  // Return early if no make flags are passed in the environment
+  const char* makeflags = getenv("MAKEFLAGS");
+  if (makeflags == nullptr) {
+    return false;
+  }
+
+  const char* jobserver_auth = "--jobserver-auth=";
+  const char* str_begin = strstr(makeflags, jobserver_auth);
```

Comment:
According to https://www.gnu.org/software/make/manual/html_node/Job-Slots.html:

```
Be aware that the MAKEFLAGS variable may contain multiple instances of the --jobserver-auth= option. Only the last instance is relevant.
```
Hence the code should probably try to reflect that.

Also it looks like there is no specific prefix on Win32 for the semaphore name (https://www.gnu.org/software/make/manual/html_node/Windows-Jobserver.html) but that `--jobserver-style=sem` should be specified in `MAKEFLAGS` as well.

I am also wary of multiple `--jobserver-style` values in the input string. I would assume that only the last one should be followed, and that `--jobserver-style=pipe` should be a condition to ignore the jobserver master.

Given this, I suggest creating a standalone static function that parses a MAKEFLAGS string value as input, and returns a struct describing the wanted jobserver type + argument (fifo path or semaphore name) from it (it doesn't have to contain Win32/Posix specific code paths), so this can be unit-tested properly with lots of edge cases.

Reply:
Good point.